src/entities/join/weighted/weightedJoin.ts

```diff
@@ -0,0 +1,208 @@
+                );
+                userData = WeightedEncoder.joinUnbalanced(maxAmountsIn, 0n);
+                break;
+            case JoinKind.SingleAsset:
+                userData = WeightedEncoder.joinSingleAsset(
+                    input.bptOut.amount,
+                    poolTokens.findIndex(
+                        (t) => t.address === input.tokenIn.toLowerCase(),
+                    ),
+                );
+                break;
+            case JoinKind.Proportional:
+                userData = WeightedEncoder.joinProportional(
+                    input.bptOut.amount,
+                );
+                break;
+        }
+
+        let tokensIn = [...poolTokens];
+        // replace wrapped token with native asset if needed
+        if (input.joinWithNativeAsset) {
+            tokensIn = poolTokens.map((token) => {
+                if (token.isUnderlyingEqual(NATIVE_ASSETS[input.chainId])) {
+                    return new Token(input.chainId, ZERO_ADDRESS, 18);
+                } else {
+                    return token;
+                }
+            });
+        }
+
+        const queryArgs = getJoinParameters({
+            poolId: poolState.id,
+            assets: tokensIn.map((t) => t.address),
+            sender: ZERO_ADDRESS,
+            recipient: ZERO_ADDRESS,
+            maxAmountsIn,
+            userData,
+        });
+
+        const client = createPublicClient({
+            transport: http(input.rpcUrl),
+            chain: CHAINS[input.chainId],
+        });
+
+        const {
+            result: [queryBptOut, queryAmountsIn],
+        } = await client.simulateContract({
+            address: BALANCER_HELPERS[input.chainId],
+            abi: balancerHelpersAbi,
+            functionName: 'queryJoin',
+            args: queryArgs,
+        });
+
+        const bpt = new Token(input.chainId, poolState.address, 18);
+        const bptOut = TokenAmount.fromRawAmount(bpt, queryBptOut);
+
+        const amountsIn = queryAmountsIn.map((a, i) =>
+            TokenAmount.fromRawAmount(tokensIn[i], a),
+        );
+
+        const tokenInIndex =
+            input.kind === JoinKind.SingleAsset
+                ? poolTokens.findIndex(
+                      (t) => t.address === input.tokenIn.toLowerCase(),
+                  )
+                : undefined;
+
+        return {
+            joinKind: input.kind,
+            id: poolState.id,
+            bptOut,
+            amountsIn,
+            tokenInIndex,
+        };
+    }
+
+    public buildCall(input: JoinCallInput): {
+        call: Address;
+        to: Address;
+        value: bigint | undefined;
+        minBptOut: bigint;
+        maxAmountsIn: bigint[];
+    } {
+        let maxAmountsIn = input.amountsIn.map((a) => a.amount);
+        let minBptOut = input.bptOut.amount;
+        let userData: Address;
+
+        switch (input.joinKind) {
+            case JoinKind.Init: {
+                maxAmountsIn = input.amountsIn.map((a) => a.amount);
```

Comment:
I find it a lot easier to reason about what `maxAmountsIn` and `minBptOut` should be for each switch case when they are both defined within the case rather than relying on previously defined default. (See [Exit](https://github.com/balancer/b-sdk/blob/exit/src/entities/exit/weighted/weightedExit.ts#L38))

Reply:
(Updated in refactor)